### PR TITLE
Refactor and fix bugs in user invitations

### PR DESCRIPTION
**Pull request by weiqingtoh**
This PR does a few things: 
 - Refactor `Course::UserInvitationService::EmailInvitationConcern` into smaller concerns. This will make it easier to read, understand and modify in the future. 
 - Strip whitespaces off emails for CSV uploads. 
 - Remove Byte Order Mark (for UTF-8) for CSVs which contain this encoding character. 
 - Use rspec's `file_fixture` for our specs. 

Solves #2744. 

This will make it easier to work on #1874 and #2841 as well. 

Suggest to review commit by commit.

### CODE REVIEW

File: app/services/concerns/course/user_invitation_service/parse_invitation_concern.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  # else it's treated like a row of student data.
+  #
+  # This method also handles the presence of UTF-8 Byte Order Marks at the
+  # start of the file, if it exists. These are invisible characters that might
+  # be persisted as the name of the student if not caught.
+  #
+  # @param [File] file Reads the given file, in CSV format, for the name and email.
+  # @return [Array<Hash>] The array of records read from the file.
+  # @raise [CSV::MalformedCSVError] When the file provided is invalid, eg. UTF-16 encoding.
+  def parse_from_file(file)
```
[the remark this answers is not in the document]
Comment (weiqingtoh) [reply app/services/concerns/course/user_invitation_service/parse_invitation_concern.rb]: Have split the function into more smaller and more readable functions.

File: app/services/concerns/course/user_invitation_service/parse_invitation_concern.rb [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  #   - Sets the name as the given email if a name was not provided.
+  #
+  # @param [Array] row Array with 3 parameters: name, email and role respectively.
+  # @return [Hash] The parsed invitation attributes given the row.
+  def parse_file_row(row)
+    return nil if row[0].blank?
+    row.unshift(row.first) unless row.length >= 2
+
+    role = parse_file_role(row[2])
+    { name: row[0], email: row[1].strip, role: role }
```
Comment (kxmbrian) [inline app/services/concerns/course/user_invitation_service/parse_invitation_concern.rb]: Are some unsuspecting users likely to do something like this? :
```
name, email, role
My Name, a@bee.co, student
```

Any harm if we also strip `name` and maybe `role`? Should we also do it for the header row to make (safely) more robust?

**Comment by kxmbrian** [comment]
Much neater now 👍

File: app/services/concerns/course/user_invitation_service/parse_invitation_concern.rb (hunk 2 above)
Comment (weiqingtoh) [reply app/services/concerns/course/user_invitation_service/parse_invitation_concern.rb]: 👍

File: app/services/course/user_invitation_service.rb [hunk 3]
```diff
@@ -13,4 +15,61 @@ def initialize(current_user, current_course)
...
+  #     Array<(Array<Course::UserInvitation>,
+  #     Array<Course::UserInvitation>,
+  #     Array<CourseUser>,
+  #     Array<CourseUser>)>
+  #   ]
+  #   A tuple containing the users newly invited, already invited,
+  #     newly registered and already registered respectively.
+  # @raise [CSV::MalformedCSVError] When the file provided is invalid.
+  def invite_users(users)
+    users = parse_invitations(users).each { |user| user[:email] = user[:email].downcase }
```
Comment (kxmbrian) [inline app/services/course/user_invitation_service.rb]: `parse_invitations` already does the downcasing

**Pull request merged**